data/abi/abi_type.go

```diff
@@ -0,0 +1,499 @@
+			parenSeg := parenSegmentRecord[parenSegCount]
+			tupleStrSegs[index] = str[parenSeg.left : parenSeg.right+1]
+			parenSegCount++
+		}
+	}
+
+	return tupleStrSegs, nil
+}
+
+// MakeUintType makes `Uint` ABI type by taking a type bitSize argument.
+// The range of type bitSize is [8, 512] and type bitSize % 8 == 0.
+func MakeUintType(typeSize uint16) (Type, error) {
+	if typeSize%8 != 0 || typeSize < 8 || typeSize > 512 {
+		return Type{}, fmt.Errorf("unsupported uint type bitSize: %d", typeSize)
+	}
+	return Type{
+		abiTypeID: Uint,
+		bitSize:   typeSize,
+	}, nil
+}
+
+// MakeByteType makes `Byte` ABI type.
+func MakeByteType() Type {
+	return Type{
+		abiTypeID: Byte,
+	}
+}
+
+// MakeUfixedType makes `UFixed` ABI type by taking type bitSize and type precision as arguments.
+// The range of type bitSize is [8, 512] and type bitSize % 8 == 0.
+// The range of type precision is [1, 160].
+func MakeUfixedType(typeSize uint16, typePrecision uint16) (Type, error) {
+	if typeSize%8 != 0 || typeSize < 8 || typeSize > 512 {
+		return Type{}, fmt.Errorf("unsupported ufixed type bitSize: %d", typeSize)
+	}
+	if typePrecision > 160 || typePrecision < 1 {
+		return Type{}, fmt.Errorf("unsupported ufixed type precision: %d", typePrecision)
+	}
+	return Type{
+		abiTypeID: Ufixed,
+		bitSize:   typeSize,
+		precision: typePrecision,
+	}, nil
+}
+
+// MakeBoolType makes `Bool` ABI type.
+func MakeBoolType() Type {
+	return Type{
+		abiTypeID: Bool,
+	}
+}
+
+// MakeStaticArrayType makes static length array ABI type by taking
+// array element type and array length as arguments.
+func MakeStaticArrayType(argumentType Type, arrayLength uint16) Type {
+	return Type{
+		abiTypeID:    ArrayStatic,
+		childTypes:   []Type{argumentType},
+		staticLength: arrayLength,
+	}
+}
+
+// MakeAddressType makes `Address` ABI type.
+func MakeAddressType() Type {
+	return Type{
+		abiTypeID: Address,
+	}
+}
+
+// MakeDynamicArrayType makes dynamic length array by taking array element type as argument.
+func MakeDynamicArrayType(argumentType Type) Type {
+	return Type{
+		abiTypeID:  ArrayDynamic,
+		childTypes: []Type{argumentType},
+	}
+}
+
+// MakeStringType makes `String` ABI type.
+func MakeStringType() Type {
+	return Type{
+		abiTypeID: String,
+	}
+}
+
+// MakeTupleType makes tuple ABI type by taking an array of tuple element types as argument.
+func MakeTupleType(argumentTypes []Type) (Type, error) {
+	if len(argumentTypes) >= (1 << 16) {
+		return Type{}, fmt.Errorf("tuple type child type number >= 2^16 error")
+	}
+	return Type{
+		abiTypeID:    Tuple,
+		childTypes:   argumentTypes,
+		staticLength: uint16(len(argumentTypes)),
+	}, nil
+}
+
+// Equal method decides the equality of two types: t == t0.
+func (t Type) Equal(t0 Type) bool {
+	if t.abiTypeID != t0.abiTypeID {
+		return false
+	} else if t.precision != t0.precision || t.bitSize != t0.bitSize {
+		return false
+	} else if t.staticLength != t0.staticLength {
+		return false
+	} else {
```

Comment:
I think these will read better without the `else`s.  Just a bunch of ifs looking like guard clauses.